src/xnumber.ts

```diff
@@ -0,0 +1,188 @@
+// =============================================================================
+// Fermat.js | Extended Number Class
+// (c) Mathigon
+// =============================================================================
+
+
+import {isInteger, nearlyEquals, numberFormat} from './arithmetic';
+import {gcd} from './number-theory';
+
+
+/**  Extended Number class. */
+export class XNumber {
+  num: number;  /** Used for all number types (decimals, fractions, units, ...). */
+  den?: number;  /** Only used for fractions and always ≥ 0. */
+
+  constructor(num: number, den?: number, public unit?: '%'|'π') {
+    // Ensure that den is always positive
+    this.num = (den !== undefined && den < 0) ? -num : num;
+    if (den !== undefined && Math.abs(den) !== 1 && num !== 0) this.den = Math.abs(den);
+  }
+
+  valueOf() {
+    return this.value;
+  }
+
+  toString() {
+    let num = `${numberFormat(this.num, 6)}`;
+    if (this.den !== undefined) num = `${num}/${numberFormat(this.den, 6)}`;
+    if (this.unit === 'π') {
+      if (num === '1') return 'π';
+      if (num === '0') return '0';
+      if (num === '–1') return '–π';
+    }
+    return `${num}${this.unit ? this.unit : ''}`;
+  }
+
+  toMathML() {
+    let str = `<mn>${this.num}</mn>`;
+    if (this.den !== undefined) str = `<mfrac>${str}<mn>${this.den}</mn></mfrac>`;
+    if (this.unit) str += (this.unit === 'π') ? `<mi>π</mi>` : `<mo>%</mo>`;
+    return str;
+  }
+
+  // ---------------------------------------------------------------------------
+
+  get value() {
```

Comment:
Some JS/TSDoc documentation here explaining that this is a conversion to decimal would be helpful to newcomers I think. (Not sure if this using `value` for this is already a convention in the codebase, though.)